File: algo_expert/recurssion/product_sum.js
// Time: O(n), n is number of ele in array
// Space: O(d), maximum depth of special array

function productSum(array, multiplier = 1) {
    let sum = 0;

    for (let ele of array) {
        if (Array.isArray(ele)) {
            sum += productSum(ele, multiplier + 1)
        } else {
            sum += ele
        }
    }
    return sum * multiplier;
}

File: leet_code/python/recurssion/product_sum.py
# O(n) time | O(d) space
def productSum(array):
    return proSumHelper(array, 1)


def proSumHelper(array, depth):
	if not array or len(array) < 1:
		return 0

	curr_sum = 0
	for elt in array:
		if type(elt) is list:
			curr_sum += proSumHelper(elt, depth+1)
		else:
			curr_sum += elt
	return curr_sum * depth
